### DISCUSSION TITLE
Will ItemsRepeater be Obsolete in Version 12?

### DISCUSSION THREAD

**Discussion by anose001** [AvaloniaUI/Avalonia#16829] [Q&A] [answered]
I am currently using `ItemsRepeater` with `UniformGridLayout` in my Avalonia production app. Someone pointed out that `ItemsRepeater` will not be available in Avalonia v12, [see comment here](https://stackoverflow.com/questions/77928874/avaloniaui-itemscontrol-virtualization#:~:text=ItemsRepeater%20is%20now,at%2019%3A31). 

Can anyone confirm or reject this claim? 

`ItemsRepeater` with `UniformGridLayout` has been working great for me for loading a very large list of images with multiple images per row where virtualization is crucial.

**Comment by maxkatz6** [comment]
See https://github.com/AvaloniaUI/Avalonia.Controls.ItemsRepeater?tab=readme-ov-file#current-status

ItemsRepeater will get its last update for 12.0 compatibility (and since we keep backward compatibility in minor updates, it will be compatible with any 12.x version as well). But project itself won't be supported anymore after 12.0, that's correct.

**Reply by maxkatz6** [reply, marked as answer]
You can keep using ItemsRepeater UniformGridLayout, as it works for you. There are also plans to implement VirtualizedUniformPanel for ItemsControl too **before** we obsolete ItemsRepeater.

**Reply by anose001** [reply]
Thanks for the quick reply. Very much appreciated.

**Reply by dbriard** [reply]
That's a bad news, I like this control a lot, especially for the Layouts. I can just grab a Layout made for WinUI and it will work in Avalonia.
ItemsControl is less convient to use as it require a VirtualizingPanel which is not easy to implement.

Also for selection, when we have ItemsRepeater of ItemsRepeaters, we can use the code of SelectionModel from WinUI for multi-level selection. ItemsControl do not allow for grouping, and multilevel selection.

It is also a lot fast when we need to scroll a large collection of items (i.e. a gallery).

The only downside of ItemsRepeater is bugs in certains conditions...

**Reply by anose001** [reply]
@maxkatz6, I assume there is a good reason to make this obsolete or maybe not? Is that the above mentioned bugs?

**Reply by maxkatz6** [reply]
@anose001 reason is simple. Instead of having two different item collection controls, and maintaining both, we chosen to only support ItemsControl as single better/stable option. Reducing confusion for newcomers and maintenance baggage for us.

**Reply by anose001** [reply]
Godda keep things slim and organized, always a good thing

**Reply by stevemonaco** [reply]
Improving existing layout containers is a good way forward. Unfortunately, there are some partial blockers like https://github.com/AvaloniaUI/Avalonia/issues/16487 for me that are demotivating. Fixing that issue may cause breaking layout changes for most Avalonia apps in existence and additional work in layout containers.

**Reply by pyking-wang** [reply]
> 您可以继续使用 ItemsRepeater UniformGridLayout，因为它适合您。我们还计划在废弃 ItemsRepeater **之前**为 ItemsControl 实现 VirtualizedUniformPanel。

Has the virtualization of ItemsControl been implemented now?

**Reply by stevemonaco** [reply]
@pyking-wang It's been there since at least 11.0. You just need to set its `ItemsPanel` to a `VirtualizingStackPanel`. The layout controls handle the virtualization of controls.

**Reply by pyking-wang** [reply]
@stevemonaco Sorry, I didn't describe it accurately. I meant VirtualizedUniformPanel.
